tools/distribution/distrotest.py

```diff
@@ -0,0 +1,715 @@
+import logging
+import os
+import re
+import shutil
+from functools import cached_property
+from itertools import chain
+from typing import Callable, Iterable, Optional, Tuple, Type, Union
+
+import verboselogs
+
+import aiodocker
+
+from tools.base import checker, utils
+from tools.docker import utils as docker_utils
+
+DISTROTEST_CONFIG_PATH = "tools/distribution/distrotest.yaml"
+
+DOCKER_IMAGE_PREFIX = "envoybuild_"
+DOCKER_CONTAINER_PREFIX = "envoytest_"
+DOCKERFILE_TEMPLATE = """
+FROM {build_image}
+{env}
+
+ADD {install_dir} {install_mount_path}
+ADD {testfile_name} {test_mount_path}
+ADD {keyfile_name} {key_mount_path}
+RUN {build_command}
+
+CMD ["tail", "-f", "/dev/null"]
+"""
+
+
+class BuildError(Exception):
+    pass
+
+
+class ConfigurationError(Exception):
+    pass
+
+
+class ContainerError(Exception):
+    pass
+
+
+class DistroTestConfig(object):
+    """Configuration object for distro tests
+
+    This holds the state/configuration across a batch of tests.
```

Comment:
hmm - checker holds the state - this holds the configuration (and shared connects which i guess is kinda state)